### DISCUSSION TITLE
Best practices to setup MFA with Supabase

### DISCUSSION THREAD

**Discussion by slengo8** [supabase/supabase#16067] [Questions]
Hey! 👋 
I have been experimenting with Supabase Auth MFA

I have went through the [docs](https://supabase.com/docs/guides/auth/auth-mfa) and I have already implemented the following flows 
- Setup MFA flow
- Login with MFA 
- Delete an MFA

I have ran into some issues and I would like to get some views on them

### Issue 1: Unverified factors are being accumulated when the user starts the MFA setup flow but do not verify the factor?
In order to show the QR code or the secret to the user I have to call the `enroll` method. This method will create a new unverified factor and if the user for some reason chose not to continue by closing the flow, we will be left with `unverified` MFA which can never be verified after we lose the QR code and the secret. This could be annoying but what is problematic is that we have an upper limit of 10 (see [docs Auth MFA](https://supabase.com/docs/reference/javascript/auth-mfa-api#:~:text=We%20don%27t%20support%20recovery%20codes%20but%20we%20allow%20users%20to%20enroll%20more%20than%201%20TOTP%20factor%2C%20with%20an%20upper%20limit%20of%2010.)) If the user reach the limit the call to enroll will fail. I can handle the error specifically and ask the user to delete unverified  factors before enrolling a new one but that is not the best UX.

My solution:
When the user cancel the MFA setup flow, I send a request to unenroll the the newly enrolled but unverified factor
The issue with this approach is that there is always the risk of request failure when we unenroll
In addition, I still show all factors in a page where the user can see and mange there factors

My question: Is there a better approach to handle this issue?

--------

### Issue 2: How can I enforce only one MFA factor?
Since only TOTP are allowed for now, I want to limit the number of TOTP for the users to be one.
Once the user setup an MFA they cannot delete it. But they can change it.

An MFA Change works as follow:
1- User verify the current MFA
2- User go through the MFA setup  flow
3- Once the new MFA is verified we unenroll the current MFA

Issue here is the risk of the failure of the unenroll request, any recommendation to handle this in a better way

----- 
### Issue 3: enrolling a Factor with a friendly name fails of there is an existing factor with the same friendly name
I want to use the friendly name but first it has to be sent with the `enroll` method which means I either have to name it myself or I have to ask the user what they want to name it before they event see it.
If I choose to populate that based on the email or the username I will get a 500 error if the friendly name is already used in another factor. 
I am not sure why the uniqueness of the friendlyName property is enforced in this way

The only solution I can think of to use the friendlyName property is to ask the user beforehand and validate it myself before sending a request 

----- 

Thank you beforehand
Looking forward to discussing these issues with you
Have a good one 😃

**Comment by harrybawsac** [comment]
I also walked into the same problem as you're having regarding issue 1.

Another possible solution would be to list the unverified factor and follow through with the MFA setup. But currently that's not possible as listing the factors doesn't return the secret or URI.

**Comment by bearbricknik** [comment]
Hi, have you found a solution to problem number 1 yet? I have the same problem and cant figure out a solution that is working reliably.

**Comment by romanroibu** [comment]
Hi all, I'm currently also trying to find a graceful way to solve _Issue 1_; what I came up with is creating an `AFTER INSERT` trigger on the `auth.mfa_factors` table that removes all the other `unverified` factors of the user:

https://github.com/romanroibu/megabase/blob/main/migrations/20240213093944_remove_other_unverified_mfa_factors.sql

Would be great to have somebody from the Supabase team to comment on this approach.

**Reply by harrybawsac** [reply]
Nice! Thanks for this.

**Reply by k-thornton** [reply]
I'm in the same boat right now as all of you.  My problem with this approach is that it doesn't guard against the case where the user starts 10 (or whatever your maximum number) enrollments without completing any of them.  They get blocked out from ever enrolling in MFA in this case.

**Reply by harrybawsac** [reply]
By the way @romanroibu I think it's easier to make it a `before insert`. Then you don't have to look for the new ID.

**Comment by Alexandre-Chapelle** [comment]
Hey, I've run into the same issue recently and I would like to share my simple solution. It's not best practice, but it works. Basically, each time a user clicks on 'enable MFA', you first remove all unverified factors and then create a new one. This approach makes it impossible to encounter the 'exceed max factors' issue mentioned by @k-thornton.

```typescript
  class AuthService {
    private supabase: SupabaseClient;

    constructor(supabaseClient: SupabaseClient) {
      this.supabase = supabaseClient;
    }

    async disableMfa(factorId: string): Promise<AuthMFAUnenrollResponse> {
      const unenroll = await this.supabase.auth.mfa.unenroll({
        factorId,
      });

      return unenroll;
    }

    async removeUnverifiedFactors() {
      const allFactors = await this.getAllMfaFactors();

      if (!allFactors.data?.all?.length) return;

      const unverifiedFactors = allFactors.data.all.filter(
        (factor) => factor.status === "unverified"
      );

      for (const factor of unverifiedFactors) {
        await this.disableMfa(factor.id);
      }
    }
    

    async getAllMfaFactors(): Promise<AuthMFAListFactorsResponse> {
      const factors = await this.supabase.auth.mfa.listFactors();
      return factors;
    }
  }
```

# Usage
And now we use it in our function
```typescript
    const handleOnEnableMfa = async () => {
      try {
        await authService.removeUnverifiedFactors();
        const enroll = await authService.enroll();
        
        // ... rest of the logic
       }
       catch (error) {
        console.error(error);}
      }
 ```

**Reply by harrybawsac** [reply]
This is a downgrade from the SQL function that was proposed, because you're making quite a lot of unneeded requests which slow things down.

1. Getting all the MFA factors is the first request.
2. Disable every single unverified one and make a request for it. You'll never know how many requests you'll make, which are not even done in parallel.
3. Only then you can start the enroll request.

I'd really advise going for something like what's already proposed.

```sql
create or replace function auth.remove_unverified_mfa_factors()
returns trigger as $$
begin
  delete from auth.mfa_factors
  where status = 'unverified'
  and user_id = new.user_id;
  return new;
end;
$$ language plpgsql security invoker;

create or replace trigger remove_unverified_mfa_factors
  before insert on auth.mfa_factors
  for each row execute function auth.remove_unverified_mfa_factors();
```

**Reply by Alexandre-Chapelle** [reply]
Hey, thanks for the detailed reply, i'll stick to your solution

**Reply by AverageCakeSlice** [reply]
@harrybawsac it looks like @Alexandre-Chapelle's solution may be the best now ever since https://github.com/orgs/supabase/discussions/34270, unless some alternative approach has been figured out since then.

**Reply by harrybawsac** [reply]
I also noticed something not working anymore. Kinda strange, because the MFA factors tables is not mentioned in that update, right?

**Reply by AverageCakeSlice** [reply]
From my understanding of the announcement, it essentially restricts manipulating almost anything on the `auth` schema, which the `mfa_factors` table lives in. There's a handful of exceptions but I don't think `mfa_tables` is included in those exceptions.

> On April 21, you will no longer be able to perform the following actions on the auth, storage, and realtime schemas:
> 
> Create tables and database functions
> Drop existing tables or database functions
> Create indexes on existing tables

**Reply by AverageCakeSlice** [reply]
I'm using Vue 3, this is a working solution that I came up with for now in my `EnrollTotp.vue` component which contains my QR code display logic:

```ts
onMounted(async () => {
    const factors = await supabase.auth.mfa.listFactors()

    if (factors.error) {
        throw factors.error
    }

    // This usually happens if the user refreshes or navigates away from the page after generating a QR code, but before verifying
    const unverifiedFactor = factors.data.all.find(
        (factor) => factor.friendly_name === TOTP_ENROLLMENT_FRIENDLY_NAME && factor.status === 'unverified'
    )

    // Supabase will automatically remove unvalidated factors after ~5 minutes, so dangling factors shouldn't be too common in practice
    if (unverifiedFactor) {
        const deleteFactorResponse = await supabase.auth.mfa.unenroll({
            factorId: unverifiedFactor.id,
        })

        if (deleteFactorResponse.error) {
            throw deleteFactorResponse.error
        }
    }

    const enrollmentCode = await supabase.auth.mfa.enroll({
        factorType: 'totp',
        friendlyName: TOTP_ENROLLMENT_FRIENDLY_NAME,
    })

    if (enrollmentCode.error) {
        throw enrollmentCode.error
    }

    factorId.value = enrollmentCode.data.id
    totpQrCode.value = enrollmentCode.data.totp.qr_code
    totpSecret.value = enrollmentCode.data.totp.secret
})
```

**Reply by AverageCakeSlice** [reply]
It does look like https://github.com/supabase/auth/pull/1371 addresses this, you just need to clear out the factors that hang out in the interim for stuff like page refreshes. But in practice with this change I don't think that filling up the `mfa_factors` table with unverified records should be a problem now that they're automatically deleted.

**Reply by AverageCakeSlice** [reply]
Also note that you don't need to use the `TOTP_ENROLLMENT_FRIENDLY_NAME` constant like I'm doing, you can just check `factor.factor_type === 'totp'` instead in your find predicate to hunt down dangling totp factors.
